[DFT][MKLGPU] Update mklgpu support to mkl preparing for oneapi 2025.0 release

s-Nick opened the pull request:
# Description

Update how dft mklgpu backend includes header files from MKL and how cmake adds them to compilation. 
This is done to avoid future name conflicts between this library and backend library. Forcing the compiler to look for some headers following the "" and <> inclusions rules accordingly with [C++ Core Guideline SF.12](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rs-incform)
# Checklist

## All Submissions

- [x] Do all unit tests pass locally? Attach a log.
[arc_mklgpu_dft_log.txt](https://github.com/user-attachments/files/17161322/arc_mklgpu_dft_log.txt)
**N.B.** Tests ran with current oneAPI release (2024.2) I cannot run tests with next one.

- [x] Have you formatted the code using clang-format?

s-Nick committed:
Update mklgpu support to mkl preparing for oneapi 2025.0 release

Update how dft mklgpu backend include header files from MKL and how
cmake add them to compilation. This is done to avoid name conflicts
between interface and library backends. Force compiler to look for some
header following the "" and <> inclusions rules accordingly with
C++ Core Guideline SF.12

s-Nick commented on src/dft/backends/mklgpu/CMakeLists.txt:
```diff
@@ -32,12 +32,14 @@ add_library(${LIB_OBJ} OBJECT
 
-target_include_directories(${LIB_OBJ}
-  PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
-)
```
Giving the includes change I made a little bit down. This command is not necessary anymore, so I decided to remove it trying to keep the cmake as lean as possible.

Rbiessy commented on src/dft/backends/mklgpu/CMakeLists.txt:
```diff
@@ -32,12 +32,14 @@ add_library(${LIB_OBJ} OBJECT
 )
+
+target_compile_options(${LIB_OBJ}
+  BEFORE PRIVATE -iquote $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
```
This deserves a comment to explain the header conflict and why the `iquote` is needed.
That looks better than I expected, this is great!

s-Nick replied:
Added comment in [562f1db](https://github.com/oneapi-src/oneMKL/pull/575/commits/562f1dbdc763fd65107eb3247b30fdfc053189b0)

Rbiessy reviewed (commented):
LGTM, thanks for this.
I'll approve once we have been able to test this with nightly builds.

s-Nick committed:
Add comment to explain why a flag is needed

s-Nick committed:
Add header file to properly check MKL version at compile-time

s-Nick committed:
Comment problematic enum for testing purposes

s-Nick committed:
Add exception check due to internal message change

The new MKL release changed the message of unimplemented exception.
Since tests were skipped by checking that message, some tests started to
fail. Update skipping check to keep compatibility with both versions.

s-Nick committed:
Remove some enum translation in mklgpu_helpers

Removing enum elements that cause compilation failure in MKL 2025.0
Apparently also previous version don't need them.

s-Nick committed:
Fix typo

s-Nick commented:
Log using next MKL 
[dft_mklgpu_2025_log.txt](https://github.com/user-attachments/files/17296156/dft_mklgpu_2025_log.txt)

lhuot commented on src/dft/backends/mklgpu/CMakeLists.txt:
```diff
@@ -32,12 +32,17 @@ add_library(${LIB_OBJ} OBJECT
 )
+
+# Due to using the same file name for different headers in this library and in
+# MKL Intel library, we force the compiler to follow C++ Core Guideline SF.12
```
```suggestion
# the Intel(R) oneAPI Math Kernel Library, we force the compiler to follow C++ Core Guideline SF.12
```

s-Nick replied:
Thank you @lhuot I updated the comment in [18bd3da](https://github.com/oneapi-src/oneMKL/pull/575/commits/18bd3da50a093f7826408b1c19384749ae93b52c)

lhuot reviewed (approved):
LGTM, minor comment added. Thank you!

s-Nick committed:
Address PR review comment

raphael-egan commented on tests/unit_tests/dft/source/compute_tests.cpp:
```diff
@@ -75,7 +75,8 @@ class ComputeTests_real_real_out_of_place_REAL
             std::string msg = e.what();                                                   \
-            if (msg.find("FFT_UNIMPLEMENTED") != std::string::npos) {                     \
+            if ((msg.find("FFT_UNIMPLEMENTED") != std::string::npos) ||                   \
+                (msg.find("Unimplemented") != std::string::npos)) {                       \
```
@s-Nick, was this addition motivated by a new exception that's _not_ a `oneapi::mkl::unimplemented` exception with "Unimplemented" in its message? If yes, could you please share details?
